### PR TITLE
Update CUDA

### PR DESCRIPTION
**Pull request by navidcy**
This PR uses the latest CUDA version which includes the fix to https://github.com/JuliaGPU/CUDA.jl/issues/1169 (see https://github.com/JuliaGPU/CUDA.jl/pull/1174)

### CODE REVIEW

**Comment by navidcy** [comment]
Seems like this bit

https://github.com/CliMA/Oceananigans.jl/blob/29b4cf9e108ea4b72004fee940c66b706b138bbb/test/test_field.jl#L269-L272

runs on the `CPU unit tests`!

https://buildkite.com/clima/oceananigans/builds/3952#f28dce5a-3fa6-486c-9bb6-a543eb6face7

Anybody has any idea why?

**Comment by navidcy** [comment]
Why not replace

https://github.com/CliMA/Oceananigans.jl/blob/29b4cf9e108ea4b72004fee940c66b706b138bbb/test/test_field.jl#L266-L272

with

```julia
for arch in archs
    ϕ = CenterField(arch, grid) 
    @test cpudata(ϕ).parent isa Array 
end
```
?

**Comment by ali-ramadhan** [comment]
Hmmm, should we be using [`CUDA.has_cuda_gpu`](https://cuda.juliagpu.org/dev/api/essentials/#CUDA.has_cuda_gpu) instead of [`CUDA.has_cuda`](https://cuda.juliagpu.org/dev/api/essentials/#CUDA.has_cuda)? Since Tartarus actually has GPUs I can see `CUDA.has_cuda_gpu` returning false (which we want) while `CUDA.has_cuda` would return true (not what we want).

This might fix the unit tests (`test_field.jl`) but there are other failures that might not be related to this.

Confusingly I see no recent changes in CUDA.jl that would lead to a difference in behavior for `has_cuda` (see changes for `src/initialization.jl`) but there were a lot of changes between v3.3.6 and v3.5.0 that I wouldn't understand: https://github.com/JuliaGPU/CUDA.jl/compare/v3.3.6...v3.5.0

**Comment by ali-ramadhan** [comment]
Yeah seems like it might be worth it to do a full search and replace, especially if tests pass right?

Since it seems that behavior has changed, I think there are some instances of `has_cuda` that should be `has_cuda_gpu` in `src/` that should also be changed.

I guess the number of users hiding their GPUs while using Oceananigans is small, but it could affect CI in the future (or one of us on Tartarus).

**Comment by glwagner** [comment]
`has_cuda_gpu` definitely sounds more accurate! Who cares about CUDA, it's the hard metal that's important.

**Comment by glwagner** [comment]
> Ah CI doesn't even initialize on the CPU now...
> 
> Kinda confused why tests were initially failing now since REPL suggests that CUDA.jl behavior actually hasn't changed (and also errors on `CUDA.has_cuda_gpu` like CI does now):
> 
> ```julia
> shell> hostname
> tartarus
> 
> julia> ENV["CUDA_VISIBLE_DEVICES"]
> "-1"
> 
> julia> using CUDA
> 
> julia> CUDA.has_cuda()
> false
> 
> julia> CUDA.has_cuda_gpu()
> ERROR: CUDA error: initialization error (code 3, ERROR_NOT_INITIALIZED)
> Stacktrace:
>  [1] throw_api_error(res::CUDA.cudaError_enum)
>    @ CUDA ~/.julia/packages/CUDA/YpW0k/lib/cudadrv/error.jl:91
>  [2] macro expansion
>    @ ~/.julia/packages/CUDA/YpW0k/lib/cudadrv/error.jl:101 [inlined]
>  [3] cuDeviceGetCount
>    @ ~/.julia/packages/CUDA/YpW0k/lib/utils/call.jl:26 [inlined]
>  [4] ndevices
>    @ ~/.julia/packages/CUDA/YpW0k/lib/cudadrv/devices.jl:160 [inlined]
>  [5] length
>    @ ~/.julia/packages/CUDA/YpW0k/lib/cudadrv/devices.jl:144 [inlined]
>  [6] has_cuda_gpu (repeats 2 times)
>    @ ~/.julia/packages/CUDA/YpW0k/src/initialization.jl:136 [inlined]
>  [7] top-level scope
>    @ REPL[5]:1
>  [8] top-level scope
>    @ ~/.julia/packages/CUDA/YpW0k/src/initialization.jl:52
> ```

err, is this a bug in `CUDA.jl`?

**Comment by simone-silvestri** [comment]
We would need CUDA 3.8.0 to develop multiGPU architectures without having to use MPI. 

You mind if I try to see what we need to do to update CUDA?

**Comment by navidcy** [comment]
@simone-silvestri if you can figure out the shenanigans with CUDA > 3.3.6 I'd be delighted! I just merged main and updated packages but I doubt that this is all what's needed.

File: benchmark/benchmark_lagrangian_particle_tracking.jl [hunk 1]
```diff
@@ -33,7 +33,7 @@ end
 
 # Benchmark parameters
 
-Architectures = has_cuda() ? [CPU, GPU] : [CPU]
+Architectures = CUDA.functional() ? [CPU, GPU] : [CPU]
```
Comment (glwagner) [inline benchmark/benchmark_lagrangian_particle_tracking.jl:36]: This same exact line appears a dozen times... perhaps time to implement a utility for it...

**Comment by francispoulin** [comment]
> We would need CUDA 3.8.0 to develop multiGPU architectures without having to use MPI.
> 
> You mind if I revamp this PR and try to see what we need to do to update CUDA? I think the tests not passing comes from the fact that we also need to update CUDAKernels to version 0.3.3 (we are now using 0.3.0)
> 
> Actually for me everything works very good from main just switching to CUDA 3.8.0 and CUDAKernels 0.3.3

That is very exciting!  Do you know of any scaling results that people have done using this multiGPU approach?  I'd be keen to try this with Oceananigans when it's working.  I know that we tried using MPI and GPU last summer and had some problems.  One, getting it to work was a pain.  Two, the effiicency dropped down to 60% when going fro 1 to 2 GPUs. I now we can do better, and maybe this is a way of doing that.

**Comment by simone-silvestri** [comment]
@navidcy I just did a lazy switch to CUDA 3.8.0 (and CUDAKernels 0.3.3) in my latest PR and it seems to be somewhat ok (for sure not all is required) but there are still some problems with `gpu_nonhydrostatic_regression` (an `device kernel invalid image`), `gpu_cubed_sphere`(an `out of bounds access`) (and for some weird reason distributed CPU??).

I ll try to figure out a bit more...

@francispoulin I don't have an indepth idea of the scaling but I with some basic trials I saw that the message passing is quite quick (the advantage is that is passes directly from GPU to GPU, which maybe is the same as cuda-aware MPI?). The advantage is surely the ease of implementing such a method

**Comment by francispoulin** [comment]
Thanks @simone-silvestri , glad to here it's easy.  I look forward to giving it a try to see how well it does in terms of scaling.

**Comment by glwagner** [comment]
> > We would need CUDA 3.8.0 to develop multiGPU architectures without having to use MPI.
> > You mind if I revamp this PR and try to see what we need to do to update CUDA? I think the tests not passing comes from the fact that we also need to update CUDAKernels to version 0.3.3 (we are now using 0.3.0)
> > Actually for me everything works very good from main just switching to CUDA 3.8.0 and CUDAKernels 0.3.3
> 
> That is very exciting! Do you know of any scaling results that people have done using this multiGPU approach? I'd be keen to try this with Oceananigans when it's working. I know that we tried using MPI and GPU last summer and had some problems. One, getting it to work was a pain. Two, the effiicency dropped down to 60% when going fro 1 to 2 GPUs. I now we can do better, and maybe this is a way of doing that.

One key problem to solve is that we cannot pass non-contiguous data between GPUs without scalar operations. This affects both "direct" communication via the CUDA.jl API and cuda-aware MPI. As a result passing non-continuous data is _extremely_ slow and is probably the primary reason for the abysmal multi-GPU scaling we saw.

I believe @simone-silvestri is going to use a single layer buffer to pass halo data (copy halo to buffer, pass to buffer on remote device, then copy from remote buffer to remote halo). Once that's implemented we can potentially use the same code for a multi-node implementation via CUDA-aware MPI.

**Comment by navidcy** [comment]
> > @simone-silvestri I opened #2255 in which I just upgraded CUDA and CUDAKernels. Feel free to work on that and make it work? Then we can close this one.
> 
> He probably wants to work on it at #2253 ?

Either! Feel free to close any PR is not needed!

**Pull request closed without merging**